tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs

```diff
@@ -0,0 +1,835 @@
+                            }
+                            break;
+                        default:
+                            throw new ArgumentException($"Unsupported method name {method}.", nameof(method));
+                    }
+                }
+
+                var coll = GetCollection(clientEncrypted, __collCollectionNamespace);
+                Insert(coll, async, corpusCopied);
+
+                var corpusDecrypted = Find(coll, new BsonDocument(), async).Single();
+                corpusDecrypted.Should().Be(corpus);
+
+                var corpusEncryptedExpected = JsonFileReader.Instance.Documents["corpus.corpus-encrypted.json"];
+                coll = GetCollection(client, __collCollectionNamespace);
+                var corpusEncryptedActual = Find(coll, new BsonDocument(), async).Single();
+                foreach (var expectedElement in corpusEncryptedExpected.Elements.Where(c => c.Value.IsBsonDocument))
+                {
+                    var expectedElementValue = expectedElement.Value;
+                    var expectedAlgorithm = ParseAlgorithm(expectedElementValue["algo"].AsString);
+                    var expectedAllowed = expectedElementValue["allowed"].ToBoolean();
+                    var expectedValue = expectedElementValue["value"];
+                    var actualValue = corpusEncryptedActual.GetValue(expectedElement.Name)["value"];
+
+                    switch (expectedAlgorithm)
+                    {
+                        case EncryptionAlgorithm.AEAD_AES_256_CBC_HMAC_SHA_512_Deterministic:
+                            actualValue.Should().Be(expectedValue);
+                            break;
+                        case EncryptionAlgorithm.AEAD_AES_256_CBC_HMAC_SHA_512_Random:
+                            if (expectedAllowed)
+                            {
+                                actualValue.Should().NotBe(expectedValue);
+                            }
+                            break;
+                        default:
+                            throw new ArgumentException($"Unsupported expected algorithm {expectedAllowed}.", nameof(expectedAlgorithm));
+                    }
+
+                    if (expectedAllowed)
+                    {
+                        var actualDecryptedValue = ExplicitDecrypt(clientEncryption, actualValue.AsBsonBinaryData, async);
+                        var expectedDecryptedValue = ExplicitDecrypt(clientEncryption, expectedValue.AsBsonBinaryData, async);
+                        actualDecryptedValue.Should().Be(expectedDecryptedValue);
+                    }
+                    else
+                    {
+                        actualValue.Should().Be(expectedValue);
+                    }
+                }
+            }
+        }
+
+        [SkippableTheory]
+        [ParameterAttributeData]
+        public void CreateDataKeyAndDoubleEncryptionTest(
+            [Values("local")] string kmsProvider,
+            [Values(false)] bool async)
```

Comment:
Shouldn't we also be testing async=true?

Reply:
done